src/mpid/ch4/netmod/ofi/ofi_init.c

```diff
@@ -1952,24 +1990,37 @@ static int addr_exchange_root_vni(MPIR_Comm * init_comm)
     int mpi_errno = MPI_SUCCESS;
     int size = MPIR_Process.size;
     int rank = MPIR_Process.rank;
+    int num_nics = MPIDI_OFI_global.num_nics;
 
     /* No pre-published address table, need do address exchange. */
-    /* First, each get its own name */
-    MPIDI_OFI_global.addrnamelen = FI_NAME_MAX;
-    MPIDI_OFI_CALL(fi_getname((fid_t) MPIDI_OFI_global.ctx[0].ep, MPIDI_OFI_global.addrname,
-                              &MPIDI_OFI_global.addrnamelen), getname);
-    MPIR_Assert(MPIDI_OFI_global.addrnamelen <= FI_NAME_MAX);
+    /* First, each process get its own name (for each nic that it will use) */
+    for (int nic = 0; nic < MPIDI_OFI_global.num_nics; nic++) {
+        MPIDI_OFI_global.addrnamelen[nic] = FI_NAME_MAX;
+        MPIDI_OFI_CALL(fi_getname((fid_t) MPIDI_OFI_global.ctx[MPIDI_OFI_get_ctx_index(0, nic)].ep,
+                                  MPIDI_OFI_global.addrname[nic],
+                                  &MPIDI_OFI_global.addrnamelen[nic]), getname);
+        MPIR_Assert(MPIDI_OFI_global.addrnamelen[nic] <= FI_NAME_MAX);
+    }
 
     /* Second, exchange names using PMI */
     /* If MPIR_CVAR_CH4_ROOTS_ONLY_PMI is true, we only collect a table of node-roots.
      * Otherwise, we collect a table of everyone. */
     void *table = NULL;
     int ret_bc_len;
-    mpi_errno = MPIDU_bc_table_create(rank, size, MPIDI_global.node_map[0],
-                                      &MPIDI_OFI_global.addrname, MPIDI_OFI_global.addrnamelen,
-                                      TRUE, MPIR_CVAR_CH4_ROOTS_ONLY_PMI, &table, &ret_bc_len);
+
+    /* Convert the 2D array to an array of pointers so the memory addresses don't get calculated
+     * incorrectly. */
+    void *addrnames[MPIDI_OFI_MAX_NICS];
+    int addrnamelens[MPIDI_OFI_MAX_NICS];
+    for (int nic = 0; nic < MPIDI_OFI_global.num_nics; nic++) {
+        addrnames[nic] = MPIDI_OFI_global.addrname[nic];
+        addrnamelens[nic] = MPIDI_OFI_global.addrnamelen[nic];
```

Comment:
If we can assume the same `addrnamelen` -- I think libfabric does that especially within the same provider, right? -- then the code can be simplified a lot. In particular, using a single 2D buffer makes sense and you can simply treat it as a single `bc`. 

But even if the addrnamelen are different, I still believe you can put all addrnames together (or concatenate them together), prepend the namelen array if necessary, and treat it as single `bc`. The code may seem more complex, but I would much prefer it since it is contained here and does not make other parts of code more complex.

Reply:
I'm not sure if you can assume that `addrnamelen` is uniform or not. I would certianly expect that within a provider, but I can't be sure.

Are you wanting to see just the local `addrname`/`addrnamelens` arrays get flattened or the globals as well? The former probably wouldn't be that big of a deal. The latter would be more ugly.

Reply:
> Are you wanting to see just the local addrname/addrnamelens arrays get flattened ?

Yes. I am glad that --
> The former probably wouldn't be that big of a deal.

Once "flattened",  the existing `MPIR_pmi_allgather_shm` will suffice, right?

Reply:
I think I remember why I didn't go this route now. If the addresses aren't all the same length, flattening them means the different processes have no way to tell the sizes of the individual entries without also having to flatten the sizes and stash them at the beginning of the entry.

When I looked at all of that, the "v" semantic started to make a lot more sense because the whole idea is that you have a bunch of entries that might have different sizes. That's why I creates `MPIR_pmi_allgatherv_shm` and used it as a replacement for `MPIR_pmi_allgather_shm` (by passing through with 1 as the number of buffers.

I also put the big comment at the top of `MPIR_pmi_allgatherv_shm` because it was confusing to me too and I wanted to try to document what I'd learned.

Reply:
I never understood how "different bc length" ever worked --
https://github.com/pmodels/mpich/blob/fae97b45655f647d3ef99037bec462b7935d29c4/src/mpid/ch4/netmod/ofi/ofi_init.c#L2014-L2017
The provider has to assume a bc length or it will have no way to guess individual bc_name in `table`, right?

Reply:
I think somewhere along the way we assume a maximum size. I'm not sure.

Reply:
> 
> 
> I think somewhere along the way we assume a maximum size. I'm not sure.

Here https://github.com/pmodels/mpich/blob/c078b1b654fe22fb2924c160905b274b6a784998/src/mpid/common/bc/mpidu_bc.c#L165

When it is not of same length, `MPIR_pmi_allgather_shm` works by assume a maximum default size. And `ucx` (which doesn't use same_lenght) works by individually creating `ep` rather than taking the whole `table`.

With `ofi`, if it is not `same_length`, and since we didn't tell the size to `fi_av_insert`, the code will not work if we ever pack the table using `MPID_MAX_BC_SIZE`.

So I believe we can take the assumption that the bc length is always the same for a single provider. Actually this is hard coded here -- https://github.com/pmodels/mpich/blob/fae97b45655f647d3ef99037bec462b7935d29c4/src/mpid/ch4/netmod/ofi/ofi_init.c#L1991-L1992

Now we agreed (I hope) that the bc length is the same, there is no objections to do local flattening before `MPIDU_bc_table_create`, right?

Reply:
Now my argument is that I'm lazy and I already wrote all the code. 😦

Reply:
I take it back, it's not just because I'm lazy. It's also because of UCX:

https://github.com/pmodels/mpich/blob/3456cfb4fceb7854595767cbc6180ca74c6a5f0f/src/mpid/ch4/netmod/ucx/ucx_init.c#L93-L96

It looks like there might need to be support for different sizes here? I'm not that familiar with what the UCX netmod is doing, but it seems like there's at least the potential for different sizes.

Reply:
UCX always has non-equal namelength, and its code doesn't depend on it. It has always been working. You can't use ucx as an excuse 😄

Reply:
Are you sure? The value of `recv_bc_len` is different than `addrname_len` and it depends on whether `same_len` is `TRUE` or `FALSE`. If we're going to remove the `same_len` argument, we'll need to add the logic in the UCX layer to deal with that.

My point isn't that these things can't be done (or just that I'm lazy 😄). It's just that in the name of "simplifying" the BC code, we add complexity elsewhere (dealing with different lengths in UCX and flattening arrays in OFI). It really seems to me like this is what these lower layers (the BC code) are for and adding a "v" function makes this a lot simpler and _doesn't_ actually make the code much more complicated because all of the code is reused between the "v" version and the regular version.

Reply:
Look, OFI code *depend* on "same length" to work. Shifting the logic to upper layer doesn't change this, except now the upper layer code is broken since the "same length" assumption is not universal there.

Reply:
We had some offline discussion and I believe we are on the same page now. As a side node and address the comment on ucx, I think we should let UCX deal with its non-equal name-lengths -- by packing into an upper-bound size. The current code does this in `src/mpid/common/bc/mpidu_bc.c` and uses `MPID_MAX_BC_SIZE`. This is not safe. Rather we should do that in `ucx_init` where we *explicitly* set a size with a comment.

Reply:
This is implemented in the version I just pushed.